hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -324,61 +343,59 @@ class Examples:
     described there.
     """
 
-    def __init__(self, record, blocks):
+    def __init__(self, record: ExampleRecord, blocks: "Blocks") -> None:
         self.trail = record.trail
         self.labels = record.labels
         self.__length = (
             self.trail.count(STOP_EXAMPLE_DISCARD_RECORD)
             + record.trail.count(STOP_EXAMPLE_NO_DISCARD_RECORD)
             + record.trail.count(DRAW_BITS_RECORD)
         )
-        self.__example_lengths = None
-
         self.blocks = blocks
-        self.__children = None
+        self.__children: "Optional[List[IntList]]" = None
 
     @calculated_example_property
     class starts_and_ends(ExampleProperty):
         def begin(self):
             self.starts = IntList.of_length(len(self.examples))
             self.ends = IntList.of_length(len(self.examples))
 
-        def start_example(self, i, label_index):
+        def start_example(self, i: int, label_index: int) -> None:
             self.starts[i] = self.bytes_read
 
-        def stop_example(self, i, label_index):
+        def stop_example(self, i: int, discarded: bool) -> None:
             self.ends[i] = self.bytes_read
 
-        def finish(self):
+        def finish(self) -> Tuple[IntList, IntList]:
             return (self.starts, self.ends)
 
     @property
-    def starts(self):
+    def starts(self) -> IntList:
         return self.starts_and_ends[0]
```

Comment:
So this part I'm a bit confused about. Looking at `ExampleProperty.finish` the base class returns an `IntList`. But this subclass (`starts_and_ends`) returns a `Tuple[IntList, IntList]`, which I think is a violation of the LSP and hence hard to type without ignoring/Any. Does this sound right or am I missing something?

Reply:
Yeah, sounds right, I'd guess we just have an expedient LSP-violating design there (and have to date gotten away with it because it's all private internals and we're just using inheritance for code sharing).